vnext/Microsoft.ReactNative/JsiWriter.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "JsiWriter.h"
+#include <crash/verifyElseCrash.h>
+
+namespace winrt::Microsoft::ReactNative {
+
+//===========================================================================
+// JsiWriter implementation
+//===========================================================================
+
+JsiWriter::JsiWriter(facebook::jsi::Runtime &runtime) noexcept : m_runtime(runtime) {
+  Push({ContinuationAction::AcceptValueAndFinish});
+}
+
+facebook::jsi::Value JsiWriter::MoveResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return std::move(m_result);
+}
+
+facebook::jsi::Value JsiWriter::CopyResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return {m_runtime, m_result};
+}
+
+void JsiWriter::WriteNull() noexcept {
+  WriteValue(facebook::jsi::Value::null());
+}
+
+void JsiWriter::WriteBoolean(bool value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteInt64(int64_t value) noexcept {
+  // JavaScript's integer is not int64_t, need to ensure that the value is in range
+  VerifyElseCrash(MinSafeInteger <= value && value <= MaxSafeInteger);
+
+  // unfortunately JSI only supports int and double for number, choose double here
```

Comment:
>supports int [](start = 28, length = 12)

Could we use int for values that fit into int?

Reply:
This is not necessary, because it becomes "number" in jsi anyway. You only read double out of int.

Reply:
It makes sense then.

---
In reply to: [430726903](https://github.com/microsoft/react-native-windows/pull/5009#discussion_r430726903) [](ancestors = 430726903)